Add NoExtendsApp rewrite

gabro opened the pull request:
Closes #174 

![image](https://user-images.githubusercontent.com/691940/26844668-806931da-4af4-11e7-879d-8f86977b8b56.png)

olafurpg commented:
Thank you for taking on this rewrite! I will try to review after the weekend :)

gabro commented:
thanks! In the meanwhile, I'll try to abstract over the template parents removal. That should really be a higher lever API in the same spirit of ImportPatchOps

xeno-by commented:
We're going to change the ASTs around parents very soon: https://github.com/scalameta/scalameta/pull/907. You may want to hold off developing this particular helper :)

gabro commented:
thanks for the heads up! I'll wait then :)
I must admit Template was in general a bit unwieldy to work with, so thanks for the work in making Trees "friendlier"

gabro commented:
@olafurpg I've pushed a working version with more test cases. I've extracted most of the utilities and added them to `ctx`. As of now I've left them in the same file, you may want to take a look and see whether it's reasonable to extract them.

gabro commented on scalafix-core/src/main/scala/scalafix/rewrite/NoExtendsApp.scala:
```diff
@@ -0,0 +1,115 @@
+        .slice(
+          ctx.tokenList.prev(from),
+          ctx.tokenList
+            .next(ctx.tokenList.next(to)) // apply next twice to include the trailing space
```
probably not the best idea...

gabro commented on scalafix-core/src/main/scala/scalafix/rewrite/NoExtendsApp.scala:
```diff
@@ -0,0 +1,115 @@
+        .asPatch
+
+    // FIXME(gabro): this is an **extremely** naive way of approaching this
+    def removeParentFromTemplate(normalized: String, template: Template)(
```
this essentially looks for the name to remove (like `App`) then check the previous keyword (`extends` or `with`). Three cases then arise:

1) `with App` => easy, just remove it
2) `extends App` => if not followed by other (i.e.` parents.length == 1`), just remove it
3) `extends App with Something` => remove `App with ` (naive but it seems to be working)

gabro replied:
not sure how solid this is, if anyone can come up with trickier test cases, it'd be nice.

gabro commented on scalafix-core/src/main/scala/scalafix/rewrite/NoExtendsApp.scala:
```diff
@@ -0,0 +1,124 @@
+        .asPatch
+    }
+
+    def removeParentFromTemplate(normalized: String, template: Template)(
```
this essentially looks for the name to remove (like `App`) then check the previous keyword (`extends` or `with`). Three cases then arise:

1) `with App` => easy, just remove it
2) `extends App` => if not followed by other (i.e.` parents.length == 1`), just remove it
3) `extends App with Something` => remove `App with ` (naive but it seems to be working)

olafurpg commented on scalafix-core/src/main/scala/scalafix/rewrite/NoExtendsApp.scala:
```diff
@@ -0,0 +1,124 @@
+}
+
+object NoExtendsAppSyntax {
+  implicit class XtensionRewriteCtx(ctx: RewriteCtx) {
```
We need a consistent story for these utilities. I'm actually thinking if we should mix them into RewriteCtx to avoid the whole extension method business. What do you think?

gabro replied:
I was thinking the same. The method extension looks unnecessarily complex and get just get away with traits.

olafurpg replied:
Great. Let's keep everything in this file then since moving things around will be a binary breaking change, I can do the migration for 0.5 https://github.com/scalacenter/scalafix/issues/193

gabro commented on scalafix-core/src/main/scala/scalafix/rewrite/NoExtendsApp.scala:
```diff
@@ -0,0 +1,124 @@
+object NoExtendsAppSyntax {
+  implicit class XtensionRewriteCtx(ctx: RewriteCtx) {
+
+    def templateBodyTokens(template: Template): Option[Tokens] = {
```
yeah that's I what I thought too after seeing the changes in that PR. It would just make the implementation simpler, but nothing hard to deal with.

gabro commented on scalafix-core/src/main/scala/scalafix/rewrite/NoExtendsApp.scala:
```diff
@@ -0,0 +1,124 @@
+      }.asPatch
+    }
+
+    def indentationTokens(tokens: Tokens): Tokens = {
```
Assuming indentation of 2 spaces seems a bit optimistic, but you've definitely seen more codebases than me ;)

olafurpg replied:
Those who are not using 2 space indent should at least be using a formatter. Scalafmt does not have an option to configure indent spaces because 2 space indent is pretty much universal.

gabro commented on scalafix-tests/input/src/main/scala/test/NoExtendsApp.scala:
```diff
@@ -0,0 +1,46 @@
+  extends Something with
+  App
+
+object Main9 extends
```
good point, I'll add a test.

gabro replied:
it fails miserably of course 😈 

I'll be working on it

olafurpg replied:
Hehe, token level rewrites are hard. I wish I could come up with a more intuitive or higher level API for common tasks

gabro replied:
done! Definitely easier than expected!

gabro commented on scalafix-tests/input/src/main/scala/test/NoExtendsApp.scala:
```diff
@@ -0,0 +1,46 @@
+/*
+rewrites = [NoExtendsApp]
```
right, I forgot about that change

gabro commented:
@olafurpg all set, ready for another review :)

gabro commented on scalafix-tests/output/src/main/scala/test/NoExtendsApp.scala:
```diff
@@ -0,0 +1,51 @@
+object Main9 extends
+  Something
+
+import scala.{ App => Ppa }
```
minor thing, in these (hopefully rare) cases, this import becomes unused. I deemed fixing this in a single pass too daunting. Users may have to manually handle this or run the `NoUnusedImports` rewrite right after.

olafurpg replied:
I wouldn't worry about that, RemoveUnusedImport also creates unused imports in cases like

```scala
import foo.bar
import bar._ // unused import

// after rewrite
import foo.bar // unused import
```

I'm afraid users will need to run scalafix at least twice to be sure all unused imports got removed.

olafurpg commented on scalafix-tests/output/src/main/scala/test/NoExtendsApp.scala:
```diff
@@ -0,0 +1,51 @@
+  }
+}
+
+object Main4 extends Something
```
Is this expected? In these cases, I'm not sure what we should do honestly. Removing App won't fix the problem however, what do you think about emitting a warning instead?

> Missing body for object that extends `scala.App`. See http://dotty.epfl.ch/docs/reference/dropped/delayed-init.html for possible workarounds.

It's possible to report warnings with `ctx.reporter.warn(pos, msg)`. There's currently no way to test reported messages.

gabro replied:
Right, I think adding a warning is our only option.
Honestly I don't know how common it is, but yes, let's warn the users.

gabro replied:
Should we remove the `App` trait anyway?

gabro commented on readme/Rewrites.scalatex:
```diff
@@ -197,5 +197,28 @@
+      }
+
+    @p
+      The @code{App} trait makes use of the @code{DelayedInit} feature which has been dropped in Dotty. While it will still be possible to use @code{App} for a concise setup of a main program, the semantics have slighly changed from Scala 2.x and it won't be possible to access the command line arguments without an explicit @code{main}.
```
I agree. The repetition is just not worth it.

gabro commented on scalafix-tests/output/src/main/scala/test/NoExtendsApp.scala:
```diff
@@ -0,0 +1,51 @@
+}
+
+object Main11 {
+  def main(args: Array[String]) = { println(args(0)) }
```
My expected output was way worse :D (I'm starting to like tokens...)

olafurpg replied:
Tokens make some things really easy and other things really hard

gabro commented on scalafix-core/src/main/scala/scalafix/rewrite/NoExtendsApp.scala:
```diff
@@ -0,0 +1,114 @@
+        .templateBodyTokens(template)
+        .map { body =>
+          val open =
+            ctx.addLeft(body.head, s"\n  def main(args: Array[String]) = {")
```
I thought about this, but I took the example verbatim from the dotty notes. Does `main` always return `Unit`?

olafurpg replied:
I see, let's still add `: Unit =`. Although that honestly just a personal preference of mine. Unit returning methods should scream "I'm doing stuff here!"

gabro replied:
Same preference. I was just concerned that `main` may return a termination value, but apparently can only be done with `sys.exit`. Moreover adding `Unit` would match the example here: https://www.scala-lang.org/documentation/getting-started.html

gabro commented:
Ready for the third pass 💃

olafurpg commented on scalafix-core/src/main/scala/scalafix/rewrite/NoExtendsApp.scala:
```diff
@@ -14,18 +14,24 @@ case class NoExtendsApp(mirror: Mirror) extends SemanticRewrite(mirror) {
+            "Missing or empty body for object that extends scala.App. See http://dotty.epfl.ch/docs/reference/dropped/delayed-init.html for possible workarounds.",
+            template.pos
+          )
+          Patch.empty
```
I've considered adding a `Message(Patch, msg: String, Severity, Position)` patch that can be used to attach a message to a patch. We could then expose an api like `ctx.warn(pos, msg) = Message(Patch.empty, msg, warn, pos)`. We would then make `ctx.reporter` private and issue the warnings at the end of the world, sorted by position location.

I think that would make more sense than the current `warn(msg): Unit` signature. What do you think?

gabro replied:
I agree it would make sense to accumulate messages, but why would they need to be attached to a `Patch`?

olafurpg replied:
You might want to attach a message to a non-empty patch to explain why a particular diff happened. This is not so useful right now since rewrites run in batch mode but it will make more sense if you want to integrate scalafix into IDEs, and create a yellow inspection message saying something like  "Hey you should not use auto-tupling, it's bad!"

gabro replied:
so we're essentially writing a linter. I'd love that, and I think we could merge the discussion into #92
